### ISSUE TITLE
[DOC/ENHANCEMENT] need to add cert-rotate feature, otherwise upgrade may stuck on Waiting for CAPI cluster fleet-local/local to be provisioned

### ISSUE THREAD

**Issue by simonflood** [harvester/harvester#3863]
**Describe the bug**

Short version: upgrade of two-node home lab cluster from 1.1.1 to 1.1.2 has got stuck at phase 3 with the apply-manifest job looping "Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...".

Longer version: I have two Intel NUC7i7DNxE NUCs as my Harvester home lab which have served me well so far, installed with 1.0.0 then upgraded through versions as available. This is the second upgrade that has failed - the first was 1.0.3 to 1.1.0 (see #3203). Clicking the Upgrade button started it doing it's thing, then when "Upgrading System Service" reached 50% with "system services" pending it stopped doing anything. Hit F5 refresh a few times but no change although at least UI is still working.

Working through the troubleshooting guide it's phase 3 where the problem appears to be. I've found various references to CAPI and fleet for upgrades to development releases of 1.1.2 but no fix.

```
dobby:~ # # phase 1
dobby:~ # kubectl get vm -n harvester-system
NAME                              AGE   STATUS    READY
upgrade-repo-hvst-upgrade-b6zlh   11d   Running   True
dobby:~ # kubectl get pods -n harvester-system | grep upgrade-repo-hvst
virt-launcher-upgrade-repo-hvst-upgrade-b6zlh-zpthk     1/1     Running     0              11d
dobby:~ # # phase 2
dobby:~ # kubectl get jobs -n cattle-system | grep prepare
No resources found in cattle-system namespace.
dobby:~ # # phase 3
dobby:~ # kubectl get jobs -n harvester-system -l harvesterhci.io/upgradeComponent=manifest
NAME                                 COMPLETIONS   DURATION   AGE
hvst-upgrade-b6zlh-apply-manifests   0/1           11d        11d
dobby:~ # kubectl logs jobs/hvst-upgrade-b6zlh-apply-manifests -n harvester-system | tail
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
dobby:~ # kubectl logs jobs/hvst-upgrade-b6zlh-apply-manifests -n harvester-system | head
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
Waiting for CAPI cluster fleet-local/local to be provisioned (current phase: Provisioning, current generation: 30259)...
```

It's now been 11 days since I started the upgrade the day 1.1.2 was released and then forgot about it!

There's nothing in the upgrade notes for 1.1.0/1.1.1 to 1.1.2 which would appear to cover this.

**To Reproduce**
1. Click upgrade button in Harvester UI to start upgrade

**Expected behavior**

The upgrade should work or if this is a known situation it should be listed in the 1.1.0/1.1.1 to 1.1.2 upgrade notes.

**Support bundle**

[supportbundle_af6572d3-42b1-43d1-8145-ce13e92f0e69_2023-05-06T18-54-35Z.zip](https://github.com/harvester/harvester/files/11413152/supportbundle_af6572d3-42b1-43d1-8145-ce13e92f0e69_2023-05-06T18-54-35Z.zip)

**Environment**
 - Harvester ISO version: 1.1.1 to 1.1.2. Originally installed with 1.0.0 then upgraded through versions as released.
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): Baremetal with Intel NUC7i7DNxE

**Additional context**
ISO was served locally from desktop via "python3 -m http.server" - I'd edited the version.yaml file to change the isoURL path to point to that before "kubectl apply -f myversion.yaml".

Except 1.0.3 to 1.1.0, previous upgrades have worked fine. Both NUCs had been powered off for a while so I connected and powered both on a few days before 1.1.2 was finally released.

**Comment by bk201** [comment]
From Rancher log, the cluster is waiting for kube-controller-manager and kube-scheduler:
```
2023/05/06 18:49:49 [INFO] [planner] rkecluster fleet-local/local: waiting: configuring bootstrap node(s) custom-24cb32ce8387: waiting for probes: kube-controller-manager, kube-scheduler
```

**Comment by himslm01** [comment]
On Slack @bk201 asked me to upload a support bundle.
Harvester created a support bundle which is too large to upload to Github, so hopefully you can download it from this link:

https://www.dropbox.com/s/r80srvdj0n0kcxa/supportbundle_5bb44244-434e-4530-ad35-35c4ef1ff661_2023-05-08T16-47-18Z.zip?dl=1

**Comment by bk201** [comment]
It could be certificate expiration: please see https://github.com/rancher/rancher/issues/41125#issuecomment-1506620040

**Comment by himslm01** [comment]
For me on the node harvester003 (which I rebooted nearly three days ago) both of those are running:
```
rancher@harvester003:~> uptime
 07:34:18  up 2 days 16:38,  1 user,  load average: 1.96, 1.87, 1.63

rancher@harvester003:~> sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock ps | grep kube-controller-manager
28e5a38a53d92       38df782a74380       2 days ago          Running             kube-controller-manager         6                   0608ea1d9c747       kube-controller-manager-harvester003

rancher@harvester003:~> sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock ps | grep kube-scheduler
2279ccabaca22       38df782a74380       2 days ago          Running             kube-scheduler                  4                   fe906f7163aa3       kube-scheduler-harvester003
```

Yes, the Harvester cluster has been up for more than a year - it was `Created: 471 days ago`.

Running the [suggested test](https://github.com/rancher/rancher/issues/41125#issuecomment-1506620040):
```
(
curl  --cacert /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.crt \
  https://127.0.0.1:10257/healthz >/dev/null 2>&1 \
  && echo "[OK] Kube Controller probe" \
  || echo "[FAIL] Kube Controller probe";

curl --cacert /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt \
  https://127.0.0.1:10259/healthz >/dev/null 2>&1  \
  && echo "[OK] Scheduler probe" \
  || echo "[FAIL] Scheduler probe";
)
```
results in:
```
[FAIL] Kube Controller probe
[FAIL] Scheduler probe
```
So I have run something similar to the suggested fix (I need sudo and the full path to crictl and the full path to the unix socket):
```
echo "Rotating kube-controller-manager certificate"
sudo rm /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.{crt,key}
sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock rm -f $(sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock ps -q --name kube-controller-manager)

echo "Rotating kube-scheduler certificate"
sudo rm /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.{crt,key}
sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock rm -f $(sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock ps -q --name kube-scheduler)
```

And the upgrade seems to have unstuck itself.

Thanks.

**Comment by w13915984028** [comment]
We are glad to see Harvester clusters are running years continuously. :)

@bk201 Has the workaround already been added to Harvester document? thanks.

**Comment by simonflood** [comment]
I can confirm my Harvester cluster had the same issue and rotating the certs allowed the upgrade to progress (after having to reboot each of my two nodes).

Note it's not the length of time the nodes have physically been up and running but the creation date of the cluster - mine was a few days before Mark.

**Comment by guangbochen** [comment]
In v1.2.0, will consider adding the doc first, once we bump the embedded Rancher to a new version with PR https://github.com/rancher/rancher/issues/41125#issuecomment-1506620040 included, the cert-rotate feature can be added on the UI.
For the upstream Rancher management, will need to add a new `cert-rotate` action and Harvester version info.
<img width="2187" alt="image" src="https://github.com/harvester/harvester/assets/4569037/141ee9e4-192b-4b7c-8e23-77958a58e658">

**Comment by bk201** [comment]
Method to rotate all nodes' service certificates:

```
kubectl edit clusters.provisioning.cattle.io local -n fleet-local
```

+=1 the generation. If the field is empty, you can use 1.
```
spec:
  ...
  rkeConfig:
    rotateCertificates:
      generation: 1
```

**Note** This needs to be done before an upgrade. If an upgrade in process, please follow the workaround in https://github.com/harvester/harvester/issues/3863#issuecomment-1539681311

**Comment by guangbochen** [comment]
The workaround doc is added in https://docs.harvesterhci.io/v1.2/upgrade/v1-1-2-to-v1-2-0/#7-upgrade-stuck-in-the-upgrading-system-service-state, move this to v1.3.0 for future enhancement, thanks.

**Comment by FrankYang0529** [comment]
For this feature, we may achieve it by following:

* Add a new setting `auto-rotate-certs-if-expiring-in-hours`. The value is a json like
```json
{
  "enable": "true",
  "hours": 480
}
```
* Watch the setting if it changes.
* Use `default/kubernetes` service to get ClusterIP.
* Use TLS connection to get certifications.
```go
    conf := &tls.Config{
        InsecureSkipVerify: true,
    }

    conn, err := tls.Dial("tcp", "<IP>:443", conf)
    if err != nil {
        log.Println("Error in Dial", err)
        return
    }
    defer conn.Close()
    certs := conn.ConnectionState().PeerCertificates
    for _, cert := range certs {
        fmt.Printf("Issuer Name: %s\n", cert.Issuer)
        fmt.Printf("Expiry: %s \n", cert.NotAfter.Format("2006-01-02T15:04:05Z07:00"))
    }
```
* If `cert.NotAfter - time.Now()` is smaller than `auto-rotate-certs-if-expiring-in-hours`, update `spec.rkeConfig.rotateCertificates.generation` in `fleet-local/local` provisioning clusters.
* If `cert.NotAfter - time.Now()` is bigger than `auto-rotate-certs-if-expiring-in-hours`, use `EnqueueAfter` to run again in the future.

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.3-df661762-head`  (02/06). 

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ UI and webhook validation on the auto-rotate-rke2-certs setting  $~~$</summary>

1. On K9s, we can found the `auto-rotate-rke2-certs` in settings

```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: harvesterhci.io/v1beta1
default: '{"enable":false,"expiringInHours":240}'
kind: Setting
metadata:
  creationTimestamp: "2024-02-16T08:14:26Z"
  generation: 1
  name: auto-rotate-rke2-certs
  resourceVersion: "6549"
  uid: 44d46a81-b947-46ba-884f-3384ac9d770d
status: {}
```


2. $\color{green}{\textsf{PASS}}$ The `enable` field must be boolean $~~$ 
  We **can't** set **none boolean** value to the `enable` field

```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# settings.harvesterhci.io "auto-rotate-rke2-certs" was not valid:
# * value: json: cannot unmarshal number into Go struct field AutoRotateRKE2Certs.enable of type bool
#
apiVersion: harvesterhci.io/v1beta1
default: '{"enable":false,"expiringInHours":240}'
...
status:
  conditions:
  - lastUpdateTime: "2024-02-16T08:19:07Z"
    status: "True"
    type: configured
value: '{"enable":123,"expiringInHours":8759}'
```


3. $\color{green}{\textsf{PASS}}$ The `expiringInHours` field must be integer and bigger or equal than 1 $~~$ 
  We **can't** set `0`, `-1` value to the `expiringInHours` field

```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# settings.harvesterhci.io "auto-rotate-rke2-certs" was not valid:
# * value: expiringInHours can't be negative or zero
#
apiVersion: harvesterhci.io/v1beta1
default: '{"enable":false,"expiringInHours":240}'
...
status:
  conditions:
  - lastUpdateTime: "2024-02-16T08:19:07Z"
    status: "True"
    type: configured
value: '{"enable":true,"expiringInHours":0}'

value: '{"enable":true,"expiringInHours":-1}'

```




 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test the expiringInHours value of auto-rotate-rke2-certs works as expected  $~~$</summary>

1. Check the certificates. find the expire date is `Feb 15 08:09:39 2025 GMT` 
```
* Server certificate:
*  subject: CN=kube-apiserver
*  start date: Feb 16 08:09:39 2024 GMT
*  expire date: Feb 15 08:09:39 2025 GMT
*  issuer: CN=rke2-server-ca@1708070979
*  SSL certificate verify result: self-signed certificate in certificate chain (19), continuing anyway.
```
2. The rkeConfig is empty in a fresh install cluster

```
harv1:~ # kubectl get clusters.provisioning.cattle.io -n fleet-local local -o yaml | yq .spec.rkeConfig
{}
```
3. Wait for 1 hour. check `rkeConfig` in fleet-local/local cluster. The `rotateCertificates` with generation: 1
```
harv1:~ # kubectl get clusters.provisioning.cattle.io -n fleet-local local -o yaml | yq .spec.rkeConfig.rotateCertificates
generation: 1
```
4. Check certificates again. the certificate expire date renewed to `Feb 18 15:15:37 2025 GMT`
```
* Server certificate:
*  subject: CN=kube-apiserver
*  start date: Feb 16 08:09:39 2024 GMT
*  expire date: Feb 18 15:15:37 2025 GMT
*  issuer: CN=rke2-server-ca@1708070979
*  SSL certificate verify result: self-signed certificate in certificate chain (19), continuing anyway.
```

 </details>

### Test Information
* Test Environment: Single and three nodes on kvm machines
* Harvester version: `v1.3-df661762-head`  (24/02/06)

### Verify Steps
<details><summary> UI and webhook validation on the auto-rotate-rke2-certs setting </summary>

1. Access to Harvester node
2. Execute command `kubectl edit settings auto-rotate-rke2-certs` 
3. Add the `value: '{"enable":true,"expiringInHours":8759}'`
4. Change the `enable` field value to non-boolean
5. Change the `expiringInHours` to out of range valeu `0`, `-1`

</details>

<details><summary> Test the expiringInHours value of auto-rotate-rke2-certs </summary>

https://github.com/harvester/harvester/pull/4985#issue-2085263525 

1. Access to Harvester node
2. Execute command `kubectl edit settings auto-rotate-rke2-certs` 
3. Add the `value: '{"enable":true,"expiringInHours":8759}'`
4. Apply a container to run curl to check certificates.
```
cat << EOF > alpine.yaml
apiVersion: v1
kind: Pod
metadata:
  name: alpine-go
  namespace: default
spec:
  containers:
  - image: golang:1.20.13-alpine3.19
    command:
      - /bin/sh
      - "-c"
      - "sleep infinity"
    imagePullPolicy: IfNotPresent
    name: alpine
  restartPolicy: Always
EOF
kubectl apply -f alpine.yaml
```
5. Install curl and check certificates. Check the certificate expire date time
```
kubectl exec -it alpine-go -- /bin/sh
/go # apk add curl
/go # curl https://<default/kubernetes service ClusterIP> -lkv
```

6. Check rkeConfig in fleet-local/local cluster. For a fresh cluster, the rkeConfig is empty.
```
kubectl get clusters.provisioning.cattle.io -n fleet-local local -o yaml | yq .spec.rkeConfig
```
7. Wait for 1 hour.
8. Check `rkeConfig` in `fleet-local/local` cluster. There is `rotateCertificates` with `generation: 1`
```
kubectl get clusters.provisioning.cattle.io -n fleet-local local -o yaml | yq .spec.rkeConfig.rotateCertificates
```
9. Check certificates again. The example expire date should be renewed.
</details>

**Comment by FrankYang0529** [comment]
Hi @TachunLin, I think the test steps you used is another issue https://github.com/harvester/harvester/issues/4992. Currently, our test plan is only test settings.value field, not settings.default. Thanks.

**Comment by TachunLin** [comment]
Thanks @FrankYang0529 for the information, 
After discussion we should add the settings value in the following format of the Harvester setting: 

```
value: '{"enable":true,"expiringInHours":8759}'
```

We would continue to retest and update the test result in the [comment](https://github.com/harvester/harvester/issues/3863#issuecomment-1930450378)

**Comment by TachunLin** [comment]
Verified **fixed** according to the test result in https://github.com/harvester/harvester/issues/3863#issuecomment-1930450378

@bk201, @FrankYang0529 
Before closing this issue, I have a question for whether we have the plan to provide the on Harvester dashboard UI settings page since currently we are adding them from k9s or kubernetes command. 
If that is the case, perhaps I can create another ticket to track the enhancement.

**Comment by w13915984028** [comment]
@FrankYang0529 The UI has this feature on master-head, maybe v1.3.0 also has it (not checked). The document needs to be updated asap, thanks.

**Comment by johnwc** [comment]
> Method to rotate all nodes' service certificates:
> 
> ```
> kubectl edit clusters.provisioning.cattle.io local -n fleet-local
> ```
> 
> +=1 the generation. If the field is empty, you can use 1.
> 
> ```
> spec:
>   ...
>   rkeConfig:
>     rotateCertificates:
>       generation: 1
> ```
> 
> **Note** This needs to be done before an upgrade. If an upgrade in process, please follow the workaround in [#3863 (comment)](https://github.com/harvester/harvester/issues/3863#issuecomment-1539681311)

@bk201 Does this make it auto rotate when the certificates get close to expiring, or does this tell it to rotate once and have to be updated manually again later to tell it to rotate it again?

**Comment by abonillabeeche** [comment]
I ran into this with 1.5.0...

**Comment by Martin-Weiss** [comment]
Running the check.sh (https://github.com/harvester/upgrade-helpers/blob/main/pre-check/v1.x/check.sh) shows me this:

```
Starting Certificates check...
/var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt has already expired.
/var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.crt has already expired.
One or more of your certificates have already expired.
Typically certs should auto-renew when the RKE service is restarted (This can also happen when you reboot a node)
You can also trigger a rotation of all a nodes' service certificates by running:
kubectl edit clusters.provisioning.cattle.io local -n fleet-local
and adding +=1 to the spec.rkeConfig.rotateCertificates.generation field or set it to 1 if it's missing.
This should be done before an upgrade. If an upgrade is already in process you might follow the workaround listed in this GitHub Issue:
https://github.com/harvester/harvester/issues/3863#issuecomment-1539681311
Certificates Test: Failed
```

--> I added the setting as requested above - rerunning the script does still show that error.. so what else is required after adding / setting that parameter?

status shows this:

```
  - lastUpdateTime: "2025-07-07T09:09:16Z"
    message: waiting for [custom-18f4ec0eb29e] certificate rotation
    reason: Waiting
    status: Unknown
    type: Provisioned
```

Workaround that I have used to continue - executed on all masters in the cluster:

```
echo "Rotating kube-controller-manager certificate"
sudo rm /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.{crt,key}
sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock rm -f $(sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock ps -q --name kube-controller-manager)

echo "Rotating kube-scheduler certificate"
sudo rm /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.{crt,key}
sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock rm -f $(sudo /var/lib/rancher/rke2/bin/crictl -r unix:///var/run/k3s/containerd/containerd.sock ps -q --name kube-scheduler)
```

**Comment by megabreit** [comment]
The issue is still present. I ran into it today with Harvester 1.5.1 and Rancher 2.11.
